### DISCUSSION TITLE
Need help to find root cause for sudden latency increase.

### DISCUSSION THREAD

**Discussion by ArenSH** [coredns/coredns#5994] [Q&A]
Hello,

Some time ago, we encountered an issue with dns resolution - latency for DNS requests increased enough to cause request timeouts on our services.

After quick investigation, we found that all CoreDNS were scheduled on the same Node. We've adjusted deployment to distribute Pods to separate Nodes and latency issue have gone. But what was the actual root cause of the problem is still unclear to me.

Our config is usual
```txt
    .:53 {
      errors
      health
      kubernetes cluster.local in-addr.arpa ip6.arpa {
        pods insecure
        fallthrough in-addr.arpa ip6.arpa
      }
      prometheus :9153
      forward . /etc/resolv.conf
      cache 60
      loop
      reload
      load
}
```

Here are some graphs from that time
The CPU usage of CoreDNS pods was relatively low
<img width="1309" alt="coredns-cpu-time" src="https://user-images.githubusercontent.com/907025/228156064-f44cfbf9-f90e-4c7e-85f8-0626c98f7d14.png">

Response latency increased to 5-25ms (on services we saw much higher latency)
<img width="1310" alt="coredns-response latency" src="https://user-images.githubusercontent.com/907025/228156813-fd65bd8b-b99a-4a86-9191-6aebde73456b.png">

From runtime metrics go's GC duration got up to 3-40ms
<img width="1311" alt="coredns-gc-duration" src="https://user-images.githubusercontent.com/907025/228157396-6b4d049d-4f19-4516-b573-0b75f6049469.png">

Goroutines number also got increased
<img width="1310" alt="coredns-goroutines-num" src="https://user-images.githubusercontent.com/907025/228157721-071a6745-e5bf-4321-8e24-c23d6491f969.png">

Upstream got couple of unusual spikes in p99.99
<img width="1308" alt="upstream latency" src="https://user-images.githubusercontent.com/907025/228183541-855d508a-ea5b-4d4d-90c6-df354522cae0.png">

All this while requests rate was relatively stable (as of coredns)
<img width="1310" alt="coredns-requests-num" src="https://user-images.githubusercontent.com/907025/228158273-d6e2a53b-5c77-483b-8e92-b862f969f78f.png">

On node where all CoreDNS pods were initially scheduled, we saw high scheduler waiting time
<img width="1305" alt="node-cpu-waiting" src="https://user-images.githubusercontent.com/907025/228158918-99b20750-28e5-4602-96de-aeb86b11df53.png">

While CPU usage was < 50%
<img width="1312" alt="cpu-usage" src="https://user-images.githubusercontent.com/907025/228161864-1e6c1da0-a311-4e4b-80bf-d2a1060f6e2d.png">

All other metrics like networking, memory, conntracks etc. were as usual, without any notable deviations.

The problem is we did not collect strace/perf/sched profilies and I cannot replicate this issue in lab environment.
I've tried to schedule many CoreDNS pods to same node (8-10 to 4 vCPU machine) to see if it is CPU contention, but it didn't look like it. Also tried to add latency to network - didn't get any increase in GC duration or goroutines num.

So I'm kinda out of any ideas what exactly caused the problematic latency increase. If you encountered a similar issue or have some thought on what else to check/investigate - please share.

**Comment by johnbelamaric** [comment]
Are you on a cloud provider or bare metal? Some cloud providers have per-node network limits unrelated to the total throughput. AWS, for example, limits network QPS based the flavor of node. Typical instances were limited at 50,000 pps as of a few years ago, as I recall. You may be hitting those limits.

**Reply by johnbelamaric** [reply]
Though that doesn't explain your charts.

**Reply by ArenSH** [reply]
Yes, we are on AWS. But network traffic is low, ~15 kpps and we should not reach any limit

<img width="1310" alt="network-pps" src="https://user-images.githubusercontent.com/907025/228273402-13249962-85a1-47ad-9c70-18b12e972856.png">

**Comment by chrisohaver** [comment]
> latency for DNS requests increased enough to cause request timeouts on our services.

In the chart, I see latency increase to 25ms ... which isn't enough latency to trigger DNS timeouts.

I presume this is the response latency reported by coredns. Do you have any "client perspective" response latency data (as measured from a client making a dns request)?  Perhaps there were network issues for dns traffic in route to/from coredns.

**Reply by ArenSH** [reply]
<img width="1307" alt="aggregated-service-latency" src="https://user-images.githubusercontent.com/907025/228270256-f09ea3f7-3e30-43ef-ad8b-429ee473413f.png">


Sorry, we have only aggregated latency metrics (cumulative duration / rps) from that time. But trend shows pretty clearly that client sees much higher latency. The "stacking" increase is due to specifics of Node.JS application - it has 4 threads for dns resolution, and if it is slow enough - outgoing requests start to pile up.

We've checked networking metrics, but there were no suspicious deviations.

**Comment by chrisohaver** [comment]
The high latency and timeouts reported by clients, vs the low CoreDNS reported latency suggests that any timeouts are caused by network delays/drops before requests enter CoreDNS (or after responses are sent to client).

That's the only conclusion I can come up with that can explain the discrepancy between coredns and client reported latency.

**Reply by ArenSH** [reply]
Sadly, we didn't observe any errors in network/udp/tcp metrics during the incident. Even infamous conntracks were fine.

I don't know how CoreDNS computes requests latency, but it can report relatively low values when, for example, packets arrived to Node, but CoreDNS reads them with delay. Or it calls some "socket write" function and go runtime takes longer time to flush it (that's all my speculations).

What is strange to me are substantially high GC duration and high waiting time for CPU - I'm not familiar with internals of CoreDNS/go and therefore don't know how to make sense of these metrics.

High waiting time and load average indicate CPU contention, but again, actual CPU usage is less than 50%.
